src/agnes.js

```diff
@@ -83,43 +73,59 @@ function centroidLink(cluster1, cluster2, disFun) {
  * @returns {number}
  */
 function wardLink(cluster1, cluster2, disFun) {
-    var x1 = 0,
-        y1 = 0,
-        x2 = 0,
-        y2 = 0;
-    for (var i = 0; i < cluster1.length; i++) {
-        x1 += cluster1[i][0];
-        y1 += cluster1[i][1];
+    return centroidLink(cluster1, cluster2, disFun)
+        *cluster1.length*cluster2.length / (cluster1.length+cluster2.length);
+}
+
+function compareNumbers(a, b) {
+    return a - b;
+}
+
+function median(values, alreadySorted) {
+    if (alreadySorted === undefined) alreadySorted = false;
+    if (!alreadySorted) {
+        values = [].concat(values).sort(compareNumbers);
     }
-    for (var j = 0; j < cluster2.length; j++) {
-        x2 += cluster2[j][0];
-        y2 += cluster2[j][1];
+    var l = values.length;
+    var half = Math.floor(l / 2);
+    if (l % 2 === 0) {
+        return (values[half - 1] + values[half]) * 0.5;
+    } else {
+        return values[half];
     }
-    x1 /= cluster1.length;
-    y1 /= cluster1.length;
-    x2 /= cluster2.length;
-    y2 /= cluster2.length;
-    return disFun([x1,y1], [x2,y2])*cluster1.length*cluster2.length / (cluster1.length+cluster2.length);
 }
 
 var defaultOptions = {
     disFunc: euclidean,
-    kind: 'single'
+    kind: 'single',
+    source:'data'
+
 };
 
 /**
  * Continuously merge nodes that have the least dissimilarity
- * @param {Array <Array <number>>} data - Array of points to be clustered
+ * @param {Array <Array <number>>} distance - Array of points to be clustered
  * @param {json} options
+ * @option source: Clustering has to be based on a list of data(default)
+ *                 or Clustering has to be based on a distance matrix(any other value)
  * @constructor
  */
 function agnes(data, options) {
-    options = options || {};
-    for (var o in defaultOptions)
-        if (!(options.hasOwnProperty(o)))
-            options[o] = defaultOptions[o];
+    options = Object.assign({}, defaultOptions, options);
     var len = data.length;
 
+    var distance = data;//If source
+    if(options.source === 'data' ) {
```

Comment:
@mljs/collaborators How to deal with the options. In this case, if "source" is set to "data", the input is an array of entries, but if "source" is set to "distance" the input is a distance matrix. But in this case, if source is something different than "data", then it is considered as a distance matrix.  Should I consider the else if(option.source === 'distance') case and throw an error if the options source is not valid??